joint_limits/src/simple_joint_limiter.cpp

```diff
@@ -0,0 +1,386 @@
+        };
+
+        // limit acc for pos_cmd and/or vel_cmd
+
+        // compute desired_acc with desired_vel and vel_state
+        desired_acc[index] =
+          (desired_vel[index] - current_joint_states.velocities[index]) / dt_seconds;
+
+        // check if decelerating - if velocity is changing toward 0
+        bool deceleration_limit_applied = false;
+        bool limit_applied = false;
+        if (
+          (desired_acc[index] < 0 && current_joint_states.velocities[index] > 0) ||
+          (desired_acc[index] > 0 && current_joint_states.velocities[index] < 0))
+        {
+          // limit deceleration
+          if (joint_limits_[index].has_deceleration_limits)
+          {
+            limit_applied = apply_acc_or_dec_limit(
+              joint_limits_[index].max_deceleration, desired_acc, limited_jnts_dec);
+            deceleration_limit_applied = true;
+          }
+        }
+
+        // limit acceleration (fallback to acceleration if no deceleration limits)
+        if (joint_limits_[index].has_acceleration_limits && !deceleration_limit_applied)
+        {
+          limit_applied = apply_acc_or_dec_limit(
+            joint_limits_[index].max_acceleration, desired_acc, limited_jnts_acc);
+        }
+
+        if (limit_applied)
+        {
+          // vel_cmd from integration of desired_acc, needed even if no vel output
+          desired_vel[index] =
+            current_joint_states.velocities[index] + desired_acc[index] * dt_seconds;
+          if (has_pos_cmd)
+          {
+            // pos_cmd from from double integration of desired_acc
+            desired_pos[index] = current_joint_states.positions[index] +
+                                 current_joint_states.velocities[index] * dt_seconds +
+                                 0.5 * desired_acc[index] * dt_seconds * dt_seconds;
+          }
+        }
+      }
+      // else we cannot compute acc, so not limiting it
+    }
+
+    // plan ahead for position limits
+    if (joint_limits_[index].has_position_limits)
+    {
+      if (has_vel_cmd && !has_pos_cmd)
+      {
+        // Check  immediate next step when using vel_cmd only, other cases already handled
+        // integrate pos
+        expected_pos[index] =
+          current_joint_states.positions[index] + desired_vel[index] * dt_seconds;
+        // if expected_pos over limit
+        auto pos = std::clamp(
+          expected_pos[index], joint_limits_[index].min_position,
+          joint_limits_[index].max_position);
+        if (pos != expected_pos[index])
+        {
+          // TODO(gwalck) compute vel_cmd that would permit to slow down in time at full
+          // deceleration in any case limit pos to max
+          expected_pos[index] = pos;
+          // and recompute vel_cmd that would lead to pos_max (not ideal as velocity would not be
+          // zero)
+          desired_vel[index] =
+            (expected_pos[index] - current_joint_states.positions[index]) / dt_seconds;
```

Comment:
Should desired acc be calculated too?

Reply:
Not necessary because it is considered later in the code. Here we expect to be close to position limits, and we will go for breaking, therefore most of the variables are called `deccel*`.